src/datachain/lib/signal_schema.py

```diff
@@ -183,37 +167,102 @@ def serialize_custom_model_fields(
+        return type_name
+
     def serialize(self) -> dict[str, Any]:
         signals: dict[str, Any] = {}
         custom_types: dict[str, Any] = {}
         for name, fr_type in self.values.items():
-            if (fr := ModelStore.to_pydantic(fr_type)) is not None:
-                ModelStore.register(fr)
-                signals[name] = ModelStore.get_name(fr)
-                type_name, fr_type = SignalSchema._get_name_original_type(fr)
-            else:
-                type_name, fr_type = SignalSchema._get_name_original_type(fr_type)
-                signals[name] = type_name
-            self.serialize_custom_model_fields(type_name, fr_type, custom_types)
+            signals[name] = self._serialize_type(fr_type, custom_types)
         if custom_types:
             signals["_custom_types"] = custom_types
         return signals
 
     @staticmethod
-    def _resolve_type(type_name: str, custom_types: dict[str, Any]) -> Optional[type]:
+    def _split_subtypes(type_name: str) -> list[str]:
+        """This splits a list of subtypes, including proper square bracket handling."""
+        start = 0
+        depth = 0
+        subtypes = []
+        for i, c in enumerate(type_name):
+            if c == "[":
+                depth += 1
+            elif c == "]":
+                if depth == 0:
+                    raise TypeError(
+                        "Extra closing square bracket when parsing subtype list"
+                    )
+                depth -= 1
+            elif c == "," and depth == 0:
+                subtypes.append(type_name[start:i].strip())
+                start = i + 1
+        if depth > 0:
+            raise TypeError("Unclosed square bracket when parsing subtype list")
+        subtypes.append(type_name[start:].strip())
+        return subtypes
+
+    @staticmethod
+    def _resolve_type(type_name: str, custom_types: dict[str, Any]) -> Optional[type]:  # noqa: PLR0911
         """Convert a string-based type back into a python type."""
+        type_name = type_name.strip()
+        if not type_name:
+            raise TypeError("Type cannot be empty")
+        if type_name == "NoneType":
+            return None
+
+        bracket_idx = type_name.find("[")
+        subtypes: Optional[tuple[Optional[type], ...]] = None
+        if bracket_idx > -1:
+            if bracket_idx == 0:
+                raise TypeError("Type cannot start with '['")
+            close_bracket_idx = type_name.rfind("]")
+            if close_bracket_idx == -1:
+                raise TypeError("Unclosed square bracket when parsing type")
+            if close_bracket_idx < bracket_idx:
+                raise TypeError("Square brackets are out of order when parsing type")
+            if close_bracket_idx == bracket_idx + 1:
+                raise TypeError("Empty square brackets when parsing type")
+            subtype_names = SignalSchema._split_subtypes(
+                type_name[bracket_idx + 1 : close_bracket_idx]
+            )
```

Comment:
I think we should investigate how others do it (eg: `pydantic`).
This looks very complicated.

Reply:
I wonder if we can use `eval` or `typing._eval_type()`/`typing.get_type_hints()`.

Reply:
I'm not aware of any pydantic-based string to type conversion, but please do mention if that is available somewhere. I did think of using `eval` but I'm not sure of the security implications / possibility for other problems from using the `eval` function. Also, this section of the code mostly just checks for syntax errors, as seen in the test `test_resolve_types_errors`.

Reply:
Aren't these trusted data (coming from `_custom_types`)?

Reply:
My two cents: https://peps.python.org/pep-0563/#resolving-type-hints-at-runtime
I’m sure you’re already aware of this PEP 👀

Reply:
> Aren't these trusted data (coming from _custom_types)?

It is also depends on where do we run this. If we run this only in DataChain (CLI or Studio workers) — it is absolutely OK to use `eval`, but if we are going to run this inside Studio backend, we should be very careful, because `_custom_types` can be easily updated by user with malformed data.

I am OK with this code, it is not covering all possible scenarios and errors might happens in case of bad data, but I think it works just fine with data generated by DataChain and I think we should not overcomplicate it.

Other option might be using regexps btw 🤔

Reply:
It's okay to `eval` even on Studio (they run inside a cluster). They are limited to types. (`typing.get_type_hints()` does exactly that).

> `_custom_types` can be easily updated by user with malformed data.

We should not allow anyone to do it through our API. But a malicious user with access to cluster can do anything they like. So that's not an issue that we need to solve (hence why we have clusters).

> but I think it works just fine with data generated by DataChain

DataChain generates arbitrary models, so it should work with anything. DataChain is no longer limited to `File` models.

We are implementing our own parser (see `_split_subtypes`) for stringized annotations. And we are adding a lot of complexity in a already complicated module.

> Other option might be using regexps btw

If we can avoid parsing this ourselves, that is my whole point.

Reply:
> It's okay to `eval` even on Studio (they run inside a cluster).

No, I mean in Studio backend, not DataChain compute cluster. And that's my only concern agains `eval`.

> We should not allow anyone to do it through our API. But a malicious user with access to cluster can do anything they like. So that's not an issue that we need to solve (hence why we have clusters).

It is OK to do anything inside the cluster, but one can send malformed data into our Studio backend and it is not OK to run `eval` there. We can not even run checks with `eval` in API endpoint because it will be run inside Studio backend, and at the same time we can not trust to checks ran in DataChain cluster.

> DataChain generates arbitrary models, so it should work with anything. DataChain is no longer limited to `File` models.
> 
> We are implementing our own parser (see `_split_subtypes`) for stringized annotations. And we are adding a lot of complexity in a already complicated module.

I agree 😥

> > Other option might be using regexps btw
> 
> If we can avoid parsing this ourselves, that is my whole point.

Yes, this is not a replacement for this parser, this is just one more way to do this, worth mentioning.

Reply:
Thank you @dtulga ❤️ For me it looks good! 👍